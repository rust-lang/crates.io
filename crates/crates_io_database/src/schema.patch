--- crates/crates_io_database/schema.rs.orig	2024-03-04 10:34:35
+++ crates/crates_io_database/schema.rs	2024-03-04 10:33:35
@@ -21,9 +21,7 @@
     /// The `pg_catalog.tsvector` SQL type
     ///
     /// (Automatically generated by Diesel.)
-    #[derive(diesel::sql_types::SqlType)]
-    #[diesel(postgres_type(name = "tsvector", schema = "pg_catalog"))]
-    pub struct Tsvector;
+    pub use diesel_full_text_search::Tsvector;
 }

 diesel::table! {
@@ -74,9 +72,9 @@
         /// (Automatically generated by Diesel.)
         revoked -> Bool,
         /// NULL or an array of crate scope patterns (see RFC #2947)
-        crate_scopes -> Nullable<Array<Nullable<Text>>>,
+        crate_scopes -> Nullable<Array<Text>>,
         /// An array of endpoint scopes or NULL for the `legacy` endpoint scope (see RFC #2947)
-        endpoint_scopes -> Nullable<Array<Nullable<Text>>>,
+        endpoint_scopes -> Nullable<Array<Text>>,
         /// The `expired_at` column of the `api_tokens` table.
         ///
         /// Its SQL type is `Nullable<Timestamptz>`.
@@ -180,12 +178,6 @@
         ///
         /// (Automatically generated by Diesel.)
         created_at -> Timestamptz,
-        /// The `path` column of the `categories` table.
-        ///
-        /// Its SQL type is `Ltree`.
-        ///
-        /// (Automatically generated by Diesel.)
-        path -> Ltree,
     }
 }

@@ -476,7 +468,7 @@
         /// Its SQL type is `Array<Nullable<Text>>`.
         ///
         /// (Automatically generated by Diesel.)
-        features -> Array<Nullable<Text>>,
+        features -> Array<Text>,
         /// The `target` column of the `dependencies` table.
         ///
         /// Its SQL type is `Nullable<Varchar>`.
@@ -703,6 +695,24 @@
 }

 diesel::table! {
+    /// Representation of the `recent_crate_downloads` view.
+    ///
+    /// This data represents the downloads in the last 90 days.
+    /// This view does not contain realtime data.
+    /// It is refreshed by the `update-downloads` script.
+    recent_crate_downloads (crate_id) {
+        /// The `crate_id` column of the `recent_crate_downloads` view.
+        ///
+        /// Its SQL type is `Integer`.
+        crate_id -> Integer,
+        /// The `downloads` column of the `recent_crate_downloads` table.
+        ///
+        /// Its SQL type is `BigInt`.
+        downloads -> BigInt,
+    }
+}
+
+diesel::table! {
     /// Representation of the `reserved_crate_names` table.
     ///
     /// (Automatically generated by Diesel.)
@@ -1018,7 +1028,8 @@
 diesel::joinable!(crate_downloads -> crates (crate_id));
 diesel::joinable!(crate_owner_invitations -> crates (crate_id));
 diesel::joinable!(crate_owners -> crates (crate_id));
-diesel::joinable!(crate_owners -> users (created_by));
+diesel::joinable!(crate_owners -> teams (owner_id));
+diesel::joinable!(crate_owners -> users (owner_id));
 diesel::joinable!(crates_categories -> categories (category_id));
 diesel::joinable!(crates_categories -> crates (crate_id));
 diesel::joinable!(crates_keywords -> crates (crate_id));
@@ -1031,6 +1042,7 @@
 diesel::joinable!(publish_limit_buckets -> users (user_id));
 diesel::joinable!(publish_rate_overrides -> users (user_id));
 diesel::joinable!(readme_renderings -> versions (version_id));
+diesel::joinable!(recent_crate_downloads -> crates (crate_id));
 diesel::joinable!(version_downloads -> versions (version_id));
 diesel::joinable!(version_owner_actions -> api_tokens (api_token_id));
 diesel::joinable!(version_owner_actions -> users (user_id));
@@ -1058,6 +1070,7 @@
     publish_limit_buckets,
     publish_rate_overrides,
     readme_renderings,
+    recent_crate_downloads,
     reserved_crate_names,
     teams,
     users,
