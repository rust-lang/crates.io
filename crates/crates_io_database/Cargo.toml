[package]
name = "crates_io_database"
version = "0.0.0"
license = "MIT OR Apache-2.0"
edition = "2021"

[lints]
workspace = true

[dependencies]
diesel = "=2.2.7"
diesel_full_text_search = "=2.2.0"

[dev-dependencies]
crates_io_test_db = { path = "../crates_io_test_db" }
diesel-async = { version = "=0.5.2", features = ["postgres"] }
tokio = { version = "=1.43.0", features = ["macros", "rt"] }
