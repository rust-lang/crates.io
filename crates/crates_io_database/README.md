# crates_io_database

This package contains the crates.io database schema as derived by `diesel print-schema`
from the database after all the migrations have been applied.

After creating new migrations (via `diesel migration generate`), you can update
the schema by running:

```sh
diesel print-schema > crates/crates_io_database/src/schema.rs
```

## `schema.patch`

Note that there is also a `schema.patch` file in this package, since the output
of `diesel-cli` needs to be tweaked a little for our purposes. For example,
it currently does not support printing materialized views in the same way as
regular tables, so we have to manually add them to the schema file.

If you need to update the patch file, you can do so by following these steps:

1. prefix `patch_file = "src/schema.patch"` in `diesel.toml` with a `#` to comment it out.
2. use `diesel print-schema` and save the output to `src/schema.rs`
3. use `cp src/schema.rs src/schema.rs.orig` to create a backup of the original file
4. use `patch src/schema.rs src/schema.patch` to apply the patch file and solve remaining issues in the `src/schema.rs` file
5. use `diff -Naur --label original --label patched src/schema.rs.orig src/schema.rs` to generate the new content for the `src/schema.patch` file
6. enable the `patch_file` option in the `diesel.toml` file again.
