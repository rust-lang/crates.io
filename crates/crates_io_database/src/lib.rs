#![doc = include_str!("../README.md")]

pub mod models;
pub mod schema;
pub mod utils;
