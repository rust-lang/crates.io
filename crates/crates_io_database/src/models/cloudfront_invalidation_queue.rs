use crate::schema::cloudfront_invalidation_queue;
use diesel::prelude::*;
use diesel_async::{AsyncPgConnection, RunQueryDsl};

#[derive(Debug, Identifiable, Queryable, QueryableByName, Selectable)]
#[diesel(table_name = cloudfront_invalidation_queue, check_for_backend(diesel::pg::Pg))]
pub struct CloudFrontInvalidationQueueItem {
    pub id: i64,
    pub path: String,
    pub created_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Insertable)]
#[diesel(table_name = cloudfront_invalidation_queue, check_for_backend(diesel::pg::Pg))]
pub struct NewCloudFrontInvalidationQueueItem<'a> {
    pub path: &'a str,
}

impl CloudFrontInvalidationQueueItem {
    /// Queue multiple invalidation paths for later processing
    pub async fn queue_paths(conn: &mut AsyncPgConnection, paths: &[String]) -> QueryResult<usize> {
        let new_items: Vec<_> = paths
            .iter()
            .map(|path| NewCloudFrontInvalidationQueueItem { path })
            .collect();

        diesel::insert_into(cloudfront_invalidation_queue::table)
            .values(&new_items)
            .execute(conn)
            .await
    }

    /// Fetch the oldest paths from the queue
    pub async fn fetch_batch(
        conn: &mut AsyncPgConnection,
        limit: i64,
    ) -> QueryResult<Vec<CloudFrontInvalidationQueueItem>> {
        // Fetch the oldest entries up to the limit
        cloudfront_invalidation_queue::table
            .order(cloudfront_invalidation_queue::created_at.asc())
            .limit(limit)
            .select(Self::as_select())
            .load(conn)
            .await
    }

    /// Remove queue items by their IDs
    pub async fn remove_items(
        conn: &mut AsyncPgConnection,
        item_ids: &[i64],
    ) -> QueryResult<usize> {
        diesel::delete(cloudfront_invalidation_queue::table)
            .filter(cloudfront_invalidation_queue::id.eq_any(item_ids))
            .execute(conn)
            .await
    }
}
