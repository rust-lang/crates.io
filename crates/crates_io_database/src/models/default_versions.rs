use crate::schema::{default_versions, versions};
use crates_io_diesel_helpers::SemverVersion;
use diesel::prelude::*;
use diesel_async::{AsyncPgConnection, RunQueryDsl};
use tracing::{debug, instrument, warn};

/// A subset of the columns of the `versions` table.
///
/// This struct is used to load all versions of a crate from the database,
/// without loading all the additional data unnecessary for default version
/// resolution.
///
/// It implements [Ord] in a way that sorts versions by the criteria specified
/// in the [update_default_version] function documentation. The default version
/// will be the "maximum" element in a sorted list of versions.
#[derive(Clone, Debug, PartialEq, Eq, Queryable, Selectable)]
#[diesel(table_name = versions)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct Version {
    pub id: i32,
    #[diesel(deserialize_as = SemverVersion)]
    pub num: semver::Version,
    pub yanked: bool,
}

impl Version {
    /// Returns `true` if the version contains a pre-release identifier.
    fn is_prerelease(&self) -> bool {
        !self.num.pre.is_empty()
    }

    fn ord_tuple(&self) -> (bool, bool, &semver::Version, i32) {
        (!self.yanked, !self.is_prerelease(), &self.num, self.id)
    }
}

impl PartialOrd for Version {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Version {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.ord_tuple().cmp(&other.ord_tuple())
    }
}

/// Updates the `default_versions` table entry for the specified crate.
///
/// This function first loads all versions of the crate from the database,
/// then determines the default version based on the following criteria:
///
/// 1. The highest non-prerelease version that is not yanked.
/// 2. The highest non-yanked version.
/// 3. The highest version.
///
/// The default version is then written to the `default_versions` table.
#[instrument(skip(conn))]
pub async fn update_default_version(
    crate_id: i32,
    conn: &mut AsyncPgConnection,
) -> QueryResult<()> {
    let default_version = calculate_default_version(crate_id, conn).await?;

    debug!(
        "Updating default version to {} (id: {})…",
        default_version.num, default_version.id
    );

    diesel::insert_into(default_versions::table)
        .values((
            default_versions::crate_id.eq(crate_id),
            default_versions::version_id.eq(default_version.id),
        ))
        .on_conflict(default_versions::crate_id)
        .do_update()
        .set(default_versions::version_id.eq(default_version.id))
        .execute(conn)
        .await?;

    Ok(())
}

/// Verifies that the default version for the specified crate is up-to-date.
#[instrument(skip(conn))]
pub async fn verify_default_version(
    crate_id: i32,
    conn: &mut AsyncPgConnection,
) -> QueryResult<()> {
    let calculated = calculate_default_version(crate_id, conn).await?;

    let saved = default_versions::table
        .select(default_versions::version_id)
        .filter(default_versions::crate_id.eq(crate_id))
        .first::<i32>(conn)
        .await
        .optional()?;

    if let Some(saved) = saved {
        if saved == calculated.id {
            debug!("Default version for crate {crate_id} is up to date");
        } else {
            warn!(
                "Default version for crate {crate_id} is outdated (expected: {saved}, actual: {})",
                calculated.id,
            );
        }
    } else {
        warn!(
            "Default version for crate {crate_id} is missing (expected: {})",
            calculated.id
        );
    }

    Ok(())
}

async fn calculate_default_version(
    crate_id: i32,
    conn: &mut AsyncPgConnection,
) -> QueryResult<Version> {
    use diesel::result::Error::NotFound;

    debug!("Loading all versions for the crate…");
    let versions = versions::table
        .filter(versions::crate_id.eq(crate_id))
        .select(Version::as_returning())
        .load::<Version>(conn)
        .await?;

    debug!("Found {} versions", versions.len());

    versions.into_iter().max().ok_or(NotFound)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::schema::crates;
    use claims::assert_some;
    use crates_io_test_db::TestDatabase;
    use insta::assert_snapshot;
    use std::fmt::Write;

    fn v(num: &str, yanked: bool) -> Version {
        let num = semver::Version::parse(num).unwrap();
        Version { id: 0, num, yanked }
    }

    #[test]
    fn test_find_default_version() {
        fn check(versions: &[Version], expected: &str) {
            let default_version = assert_some!(versions.iter().max());
            assert_eq!(default_version.num.to_string(), expected);
        }

        // Only a single version
        let versions = vec![v("1.0.0", false)];
        check(&versions, "1.0.0");

        // Multiple versions out of order
        let versions = vec![
            v("1.0.0", false),
            v("1.0.1", false),
            v("1.1.0", false),
            v("1.0.2", false),
        ];
        check(&versions, "1.1.0");

        // Multiple versions with one pre-release
        let versions = vec![
            v("1.0.0", false),
            v("1.1.0", false),
            v("2.0.0-beta.1", false),
        ];
        check(&versions, "1.1.0");

        // Only pre-release versions
        let versions = vec![
            v("1.0.0-beta.1", false),
            v("1.0.0-beta.2", false),
            v("1.0.0-beta.3", false),
        ];
        check(&versions, "1.0.0-beta.3");

        // Only pre-release versions, with highest yanked
        let versions = vec![
            v("1.0.0-beta.1", false),
            v("1.0.0-beta.2", false),
            v("1.0.0-beta.3", true),
        ];
        check(&versions, "1.0.0-beta.2");

        // Only yanked versions
        let versions = vec![
            v("1.0.0-beta.1", true),
            v("1.0.0-beta.2", true),
            v("1.0.0-beta.3", true),
        ];
        check(&versions, "1.0.0-beta.3");
    }

    #[test]
    fn test_ord() {
        let mut versions = vec![
            v("1.0.0", false),
            v("1.0.0-beta.1", false),
            v("1.0.0-beta.2", false),
            v("1.0.0-beta.3", false),
            v("1.0.1", true),
            v("1.0.2", false),
            v("1.1.0", false),
            v("1.1.1-beta.1", true),
            v("1.1.1", true),
            v("1.0.3", false),
            v("2.0.0-beta.1", false),
        ];

        versions.sort();

        assert_snapshot!(format_versions(&versions), @r"
        1.1.1-beta.1 (yanked)
        1.0.1 (yanked)
        1.1.1 (yanked)
        1.0.0-beta.1
        1.0.0-beta.2
        1.0.0-beta.3
        2.0.0-beta.1
        1.0.0
        1.0.2
        1.0.3
        1.1.0
        ");
    }

    fn format_versions(versions: &[Version]) -> String {
        let mut buf = String::with_capacity(versions.len() * 20);
        for v in versions {
            write!(buf, "{}", v.num).unwrap();
            if v.yanked {
                buf.push_str(" (yanked)");
            }
            buf.push('\n');
        }
        buf
    }

    async fn create_crate(name: &str, conn: &mut AsyncPgConnection) -> i32 {
        diesel::insert_into(crates::table)
            .values(crates::name.eq(name))
            .returning(crates::id)
            .get_result(conn)
            .await
            .unwrap()
    }

    async fn create_version(crate_id: i32, num: &str, conn: &mut AsyncPgConnection) {
        diesel::insert_into(versions::table)
            .values((
                versions::crate_id.eq(crate_id),
                versions::num.eq(num),
                versions::num_no_build.eq(num),
                versions::checksum.eq(""),
                versions::crate_size.eq(0),
            ))
            .execute(conn)
            .await
            .unwrap();
    }

    async fn get_default_version(crate_id: i32, conn: &mut AsyncPgConnection) -> String {
        default_versions::table
            .inner_join(versions::table)
            .select(versions::num)
            .filter(default_versions::crate_id.eq(crate_id))
            .first(conn)
            .await
            .unwrap()
    }

    #[tokio::test]
    async fn test_update_default_version() {
        let test_db = TestDatabase::new();
        let conn = &mut test_db.async_connect().await;

        let crate_id = create_crate("foo", conn).await;
        create_version(crate_id, "1.0.0", conn).await;

        update_default_version(crate_id, conn).await.unwrap();
        assert_eq!(get_default_version(crate_id, conn).await, "1.0.0");

        create_version(crate_id, "1.1.0", conn).await;
        create_version(crate_id, "1.0.1", conn).await;
        assert_eq!(get_default_version(crate_id, conn).await, "1.0.0");

        update_default_version(crate_id, conn).await.unwrap();
        assert_eq!(get_default_version(crate_id, conn).await, "1.1.0");
    }
}
