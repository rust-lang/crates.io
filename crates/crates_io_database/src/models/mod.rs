pub use self::action::{NewVersionOwnerAction, VersionAction, VersionOwnerAction};
pub use self::category::{Category, CrateCategory, NewCategory};
pub use self::cloudfront_invalidation_queue::CloudFrontInvalidationQueueItem;
pub use self::crate_owner_invitation::{
    CrateOwnerInvitation, NewCrateOwnerInvitation, NewCrateOwnerInvitationOutcome,
};
pub use self::default_versions::{update_default_version, verify_default_version};
pub use self::deleted_crate::NewDeletedCrate;
pub use self::dependency::{Dependency, DependencyKind, ReverseDependency};
pub use self::download::VersionDownload;
pub use self::email::{Email, NewEmail};
pub use self::follow::Follow;
pub use self::keyword::{CrateKeyword, Keyword};
pub use self::krate::{Crate, CrateName, NewCrate};
pub use self::owner::{CrateOwner, Owner, OwnerKind};
pub use self::team::{NewTeam, Team};
pub use self::token::ApiToken;
pub use self::trustpub::TrustpubData;
pub use self::user::{NewUser, User};
pub use self::version::{NewVersion, TopVersions, Version};

pub mod helpers;

mod action;
pub mod category;
mod cloudfront_invalidation_queue;
pub mod crate_owner_invitation;
pub mod default_versions;
mod deleted_crate;
pub mod dependency;
pub mod download;
mod email;
mod follow;
mod keyword;
pub mod krate;
mod owner;
pub mod team;
pub mod token;
pub mod trustpub;
pub mod user;
pub mod version;
pub mod versions_published_by;
