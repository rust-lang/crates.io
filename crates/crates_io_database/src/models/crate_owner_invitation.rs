use chrono::{DateTime, Utc};
use diesel::prelude::*;
use diesel_async::scoped_futures::ScopedFutureExt;
use diesel_async::{AsyncConnection, AsyncPgConnection, RunQueryDsl};
use secrecy::SecretString;

use crate::models::CrateOwner;
use crate::schema::{crate_owner_invitations, crates};

#[derive(Debug)]
pub enum NewCrateOwnerInvitationOutcome {
    AlreadyExists,
    InviteCreated { plaintext_token: SecretString },
}

#[derive(Clone, Debug, Insertable)]
#[diesel(table_name = crate_owner_invitations, check_for_backend(diesel::pg::Pg))]
pub struct NewCrateOwnerInvitation {
    pub invited_user_id: i32,
    pub invited_by_user_id: i32,
    pub crate_id: i32,
    pub expires_at: DateTime<Utc>,
}

impl NewCrateOwnerInvitation {
    pub async fn create(
        &self,
        conn: &mut AsyncPgConnection,
    ) -> QueryResult<NewCrateOwnerInvitationOutcome> {
        // Before actually creating the invite, check if an expired invitation already exists
        // and delete it from the database. This allows obtaining a new invite if the old one
        // expired, instead of returning "already exists".
        diesel::delete(crate_owner_invitations::table)
            .filter(crate_owner_invitations::invited_user_id.eq(self.invited_user_id))
            .filter(crate_owner_invitations::crate_id.eq(self.crate_id))
            .filter(crate_owner_invitations::expires_at.le(Utc::now()))
            .execute(conn)
            .await?;

        let res: Option<CrateOwnerInvitation> = diesel::insert_into(crate_owner_invitations::table)
            .values(self)
            // The ON CONFLICT DO NOTHING clause results in not creating the invite if another one
            // already exists. This does not cause problems with expired invitation as those are
            // deleted before doing this INSERT.
            .on_conflict_do_nothing()
            .get_result(conn)
            .await
            .optional()?;

        Ok(match res {
            Some(record) => NewCrateOwnerInvitationOutcome::InviteCreated {
                plaintext_token: record.token,
            },
            None => NewCrateOwnerInvitationOutcome::AlreadyExists,
        })
    }
}

/// The model representing a row in the `crate_owner_invitations` database table.
#[derive(Clone, Debug, Identifiable, Queryable)]
#[diesel(primary_key(invited_user_id, crate_id))]
pub struct CrateOwnerInvitation {
    pub invited_user_id: i32,
    pub invited_by_user_id: i32,
    pub crate_id: i32,
    pub created_at: DateTime<Utc>,
    #[diesel(deserialize_as = String)]
    pub token: SecretString,
    pub expires_at: DateTime<Utc>,
}

impl CrateOwnerInvitation {
    pub async fn find_by_id(
        user_id: i32,
        crate_id: i32,
        conn: &mut AsyncPgConnection,
    ) -> QueryResult<Self> {
        crate_owner_invitations::table
            .find((user_id, crate_id))
            .first::<Self>(conn)
            .await
    }

    pub async fn find_by_token(token: &str, conn: &mut AsyncPgConnection) -> QueryResult<Self> {
        crate_owner_invitations::table
            .filter(crate_owner_invitations::token.eq(token))
            .first::<Self>(conn)
            .await
    }

    pub async fn accept(self, conn: &mut AsyncPgConnection) -> Result<(), AcceptError> {
        if self.is_expired() {
            let crate_name: String = crates::table
                .find(self.crate_id)
                .select(crates::name)
                .first(conn)
                .await?;

            return Err(AcceptError::Expired { crate_name });
        }

        conn.transaction(|conn| {
            async move {
                CrateOwner::from_invite(&self).insert(conn).await?;

                diesel::delete(&self).execute(conn).await?;

                Ok(())
            }
            .scope_boxed()
        })
        .await
    }

    pub async fn decline(self, conn: &mut AsyncPgConnection) -> QueryResult<()> {
        // The check to prevent declining expired invitations is *explicitly* missing. We do not
        // care if an expired invitation is declined, as that just removes the invitation from the
        // database.

        diesel::delete(&self).execute(conn).await?;
        Ok(())
    }

    pub fn is_expired(&self) -> bool {
        self.expires_at <= Utc::now()
    }
}

#[derive(Debug, thiserror::Error)]
pub enum AcceptError {
    #[error(transparent)]
    Diesel(#[from] diesel::result::Error),
    #[error("The invitation has expired")]
    Expired { crate_name: String },
}
