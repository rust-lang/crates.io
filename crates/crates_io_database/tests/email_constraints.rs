//! Tests to verify that the SQL constraints on the `emails` table are enforced correctly.

use crates_io_database::models::{Email, NewEmail, NewUser};
use crates_io_database::schema::{emails, users};
use crates_io_test_db::TestDatabase;
use diesel::prelude::*;
use diesel::result::Error;
use diesel_async::{AsyncPgConnection, RunQueryDsl};
use insta::assert_debug_snapshot;

const MAX_EMAIL_COUNT: i32 = 32;

#[derive(Debug)]
#[allow(dead_code)]
/// A snapshot of the email data used for testing.
/// This struct is used to compare the results of database operations against expected values.
/// We can't use `Email` directly because it contains date/time fields that would change each time.
struct EmailSnapshot {
    id: i32,
    user_id: i32,
    email: String,
    primary: bool,
}
impl From<Email> for EmailSnapshot {
    fn from(email: Email) -> Self {
        EmailSnapshot {
            id: email.id,
            user_id: email.user_id,
            email: email.email,
            primary: email.primary,
        }
    }
}

// Insert a test user into the database and return its ID.
async fn insert_test_user(conn: &mut AsyncPgConnection) -> i32 {
    let user_count = users::table.count().get_result::<i64>(conn).await.unwrap();
    let user = NewUser::builder()
        .name(&format!("testuser{}", user_count + 1))
        .gh_id(user_count as i32 + 1)
        .gh_login(&format!("testuser{}", user_count + 1))
        .gh_access_token("token")
        .build()
        .insert(conn)
        .await
        .unwrap();
    user.id
}

// Insert a basic primary email for a user.
async fn insert_static_primary_email(
    conn: &mut AsyncPgConnection,
    user_id: i32,
) -> Result<Email, Error> {
    NewEmail::builder()
        .user_id(user_id)
        .email("primary@example.com")
        .primary(true)
        .build()
        .insert(conn)
        .await
}

// Insert a basic secondary email for a user.
async fn insert_static_secondary_email(
    conn: &mut AsyncPgConnection,
    user_id: i32,
) -> Result<Email, Error> {
    NewEmail::builder()
        .user_id(user_id)
        .email("secondary@example.com")
        .primary(false)
        .build()
        .insert(conn)
        .await
}

#[tokio::test]
// Add a primary email address to the database.
async fn create_primary_email() {
    let test_db = TestDatabase::new();
    let mut conn = test_db.async_connect().await;
    let user_id: i32 = insert_test_user(&mut conn).await;

    let result = insert_static_primary_email(&mut conn, user_id)
        .await
        .map(EmailSnapshot::from);

    assert_debug_snapshot!(result);
}

#[tokio::test]
// Attempt to create a secondary email address without a primary already present, which should fail.
// This tests the `verify_exactly_one_primary_email` trigger.
async fn create_secondary_email_without_primary() {
    let test_db = TestDatabase::new();
    let mut conn = test_db.async_connect().await;
    let user_id: i32 = insert_test_user(&mut conn).await;

    let result = insert_static_secondary_email(&mut conn, user_id).await;

    assert_debug_snapshot!(result);
}

#[tokio::test]
// Attempt to delete the only email address for a user, which should succeed.
// This tests the `prevent_primary_email_deletion` trigger.
async fn delete_only_email() {
    let test_db = TestDatabase::new();
    let mut conn = test_db.async_connect().await;
    let user_id: i32 = insert_test_user(&mut conn).await;

    let email = insert_static_primary_email(&mut conn, user_id)
        .await
        .expect("failed to insert primary email");

    let result = diesel::delete(emails::table.find(email.id))
        .execute(&mut conn)
        .await;

    assert_debug_snapshot!(result);
}

#[tokio::test]
// Add a secondary email address to the database.
async fn create_secondary_email() {
    let test_db = TestDatabase::new();
    let mut conn = test_db.async_connect().await;
    let user_id: i32 = insert_test_user(&mut conn).await;

    let primary = insert_static_primary_email(&mut conn, user_id)
        .await
        .map(EmailSnapshot::from);

    let secondary = insert_static_secondary_email(&mut conn, user_id)
        .await
        .map(EmailSnapshot::from);

    assert_debug_snapshot!((primary, secondary));
}

#[tokio::test]
// Attempt to delete a secondary email address, which should succeed.
// This tests the `prevent_primary_email_deletion` trigger.
async fn delete_secondary_email() {
    let test_db = TestDatabase::new();
    let mut conn = test_db.async_connect().await;
    let user_id: i32 = insert_test_user(&mut conn).await;

    let _primary = insert_static_primary_email(&mut conn, user_id)
        .await
        .expect("failed to insert primary email");

    let secondary = insert_static_secondary_email(&mut conn, user_id)
        .await
        .expect("failed to insert secondary email");

    let result = diesel::delete(emails::table.find(secondary.id))
        .execute(&mut conn)
        .await;

    assert_debug_snapshot!(result);
}

#[tokio::test]
// Attempt to delete a primary email address when a secondary email exists, which should fail.
// This tests the `prevent_primary_email_deletion` trigger.
async fn delete_primary_email_with_secondary() {
    let test_db = TestDatabase::new();
    let mut conn = test_db.async_connect().await;
    let user_id: i32 = insert_test_user(&mut conn).await;

    let primary = insert_static_primary_email(&mut conn, user_id)
        .await
        .expect("failed to insert primary email");

    let _secondary = insert_static_secondary_email(&mut conn, user_id)
        .await
        .expect("failed to insert secondary email");

    let result = diesel::delete(emails::table.find(primary.id))
        .execute(&mut conn)
        .await;

    assert_debug_snapshot!(result);
}

#[tokio::test]
// Attempt to add a secondary email address with the same email as the primary, which should fail.
// This tests the `unique_user_email` constraint.
async fn create_secondary_email_with_same_email_as_primary() {
    let test_db = TestDatabase::new();
    let mut conn = test_db.async_connect().await;
    let user_id: i32 = insert_test_user(&mut conn).await;

    let primary = insert_static_primary_email(&mut conn, user_id)
        .await
        .map(EmailSnapshot::from)
        .expect("failed to insert primary email");

    let secondary = NewEmail::builder()
        .user_id(user_id)
        .email(&primary.email)
        .primary(false)
        .build()
        .insert(&mut conn)
        .await
        .map(EmailSnapshot::from);

    assert_debug_snapshot!((primary, secondary));
}

#[tokio::test]
// Attempt to create more than the maximum allowed emails for a user, which should fail.
// This tests the `enforce_max_emails_per_user` trigger.
async fn create_too_many_emails() {
    let test_db = TestDatabase::new();
    let mut conn = test_db.async_connect().await;
    let user_id: i32 = insert_test_user(&mut conn).await;

    let mut errors = Vec::new();
    for i in 0..MAX_EMAIL_COUNT + 2 {
        let result = NewEmail::builder()
            .user_id(user_id)
            .email(&format!("me+{i}@example.com"))
            .primary(i == 0)
            .build()
            .insert(&mut conn)
            .await
            .map(EmailSnapshot::from);

        if let Err(err) = result {
            errors.push(err);
        }
    }

    assert_debug_snapshot!(errors);
}

#[tokio::test]
// Attempt to add the same email address to two users, which should succeed.
async fn create_same_email_for_different_users() {
    let test_db = TestDatabase::new();
    let mut conn = test_db.async_connect().await;

    let user_id_1: i32 = insert_test_user(&mut conn).await;
    let user_id_2: i32 = insert_test_user(&mut conn).await;

    let first = insert_static_primary_email(&mut conn, user_id_1)
        .await
        .map(EmailSnapshot::from);

    let second = insert_static_primary_email(&mut conn, user_id_2)
        .await
        .map(EmailSnapshot::from);

    assert_debug_snapshot!((first, second));
}

#[tokio::test]
// Create a primary email, a secondary email, and then promote the secondary email to primary.
// This tests the `promote_email_to_primary` function and the `unique_primary_email_per_user` constraint.
async fn promote_secondary_to_primary() {
    let test_db = TestDatabase::new();
    let mut conn = test_db.async_connect().await;
    let user_id: i32 = insert_test_user(&mut conn).await;

    let _primary = insert_static_primary_email(&mut conn, user_id)
        .await
        .expect("failed to insert primary email");

    let secondary = insert_static_secondary_email(&mut conn, user_id)
        .await
        .expect("failed to insert secondary email");

    diesel::sql_query("SELECT promote_email_to_primary($1)")
        .bind::<diesel::sql_types::Integer, _>(secondary.id)
        .execute(&mut conn)
        .await
        .expect("failed to promote secondary email to primary");

    // Query both emails to verify that the primary flag has been updated correctly for both.
    let result = emails::table
        .select((emails::id, emails::email, emails::primary))
        .load::<(i32, String, bool)>(&mut conn)
        .await;

    assert_debug_snapshot!(result);
}

#[tokio::test]
// Attempt to demote a primary email to secondary without assigning another primary, which should fail.
// This tests the `verify_exactly_one_primary_email` trigger.
async fn demote_primary_without_new_primary() {
    let test_db = TestDatabase::new();
    let mut conn = test_db.async_connect().await;
    let user_id: i32 = insert_test_user(&mut conn).await;

    let primary = insert_static_primary_email(&mut conn, user_id)
        .await
        .expect("failed to insert primary email");

    let result = diesel::update(emails::table.find(primary.id))
        .set(emails::primary.eq(false))
        .execute(&mut conn)
        .await;

    assert_debug_snapshot!(result);
}

#[tokio::test]
// Attempt to create a primary email when one already exists for the user, which should fail.
// This tests the `unique_primary_email_per_user` constraint.
async fn create_primary_email_when_one_exists() {
    let test_db = TestDatabase::new();
    let mut conn = test_db.async_connect().await;
    let user_id: i32 = insert_test_user(&mut conn).await;

    let first = insert_static_primary_email(&mut conn, user_id)
        .await
        .map(EmailSnapshot::from);

    let second = NewEmail::builder()
        .user_id(user_id)
        .email("me+2@example.com")
        .primary(true)
        .build()
        .insert(&mut conn)
        .await
        .map(EmailSnapshot::from);

    assert_debug_snapshot!((first, second));
}
