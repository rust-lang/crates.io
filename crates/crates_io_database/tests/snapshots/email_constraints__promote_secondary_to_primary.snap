---
source: crates/crates_io_database/tests/email_constraints.rs
expression: result
---
Ok(
    [
        (
            1,
            "primary@example.com",
            false,
        ),
        (
            2,
            "secondary@example.com",
            true,
        ),
    ],
)
